### PR TITLE
Bug - Search page queries

### LINKED ISSUE #4557: Bug - Search page need to search by available candidates using pool_candidate_status
(UPDATED)

## Description
After [further investigation from Peter](https://github.com/GCTC-NTGC/gc-digital-talent/issues/4557#issuecomment-1301032528) (thanks Peter😄), it looks like the problem was not with the Operational Requirements filter, but from an  inconsistency btw the filter dialog and the search page. 

`The search page is scoping candidates by (user) job_looking_status in ('ACTIVELY_LOOKING','OPEN_TO_OPPORTUNITIES') while the filter dialog is scoping candidates by "pool_candidate_status" in ('PLACED_CASUAL','QUALIFIED_AVAILABLE')`

Fix this problem by setting the search page be fixed to search by pool_candidate_status" in ('PLACED_CASUAL','QUALIFIED_AVAILABLE') for the release.

## Acceptance criteria
- [ ] Agreed, Search page needs to be fixed to search by pool_candidate_status" in ('PLACED_CASUAL','QUALIFIED_AVAILABLE') for the release.
- [ ] Double check that countapplicants is only returning users with that filter

### PR DESCRIPTION
**Pull request by yonikid15**
Resolves #4557 

## Notes
- Alright, I believe the solution was filtering by the [available users in pools](https://github.com/GCTC-NTGC/gc-digital-talent/blob/bug/4557-search-page-queries/api/app/Models/User.php#L260-L283) in the CountPoolCandidatesByPool query builder. This ensures the pool candidates are QUALIFIED_AVAILABLE or PLACED_CASUAL, and ensures the pool candidate hasn't expired. 
```php
    public static function filterByAvailableInPools(Builder $query, ?array $poolIds): Builder
    {
        if (empty($poolIds)) {
            return $query;
        }
        $poolFilters = [];
        foreach ($poolIds as $index => $poolId) {
            $poolFilters[$index] = [
                'poolId' => $poolId,
                'expiryStatus' => ApiEnums::CANDIDATE_EXPIRY_FILTER_ACTIVE,
                'statuses' => [ApiEnums::CANDIDATE_STATUS_QUALIFIED_AVAILABLE, ApiEnums::CANDIDATE_STATUS_PLACED_CASUAL]
            ];
        }
        return self::filterByPools($query, $poolFilters);
    }
```

## Testing
- Go to [search page](http://localhost:8000/en/search) and fill out filters. View TOTAL number of candidates, and the number of candidates in each pool. The TOTAL number should never be greater than the sum of all the pool card totals. This is because one applicant can be found in multiple pools.
- Submit a request for candidates.
- Next, on the admin side check to see if the number of candidates is also correct. 
- Go to Requests table. Then click the request you created, scroll down to the table, and ensure that the correct number of candidates show up.

Please feel free to reach out to me for any help testing 🧪

### CODE REVIEW

**Comment by vd1992** [comment]
playing around with the failing test
first this change is needed, otherwise the randomly generated pool_candidate_status messes with things, need to restrict the status to QUALIFIED_AVAILABLE or PLACED_CASUAL

image

**Review by vd1992 — CHANGES_REQUESTED** [review]
the filtering by pools doesn't seem to work anymore, it returns all

image

**Comment by vd1992** [comment]
I can see that the shape of the input that goes into `User::filterByAvailableInPools()` is different from what goes into `PoolCandidate::filterByPools`

The former takes an an array of ids seen here

image

Whereas the latter takes an array of pool objects and pulls id from each object seen here

image

even accounting for that I still can't figure out why it isn't working 😵 

### CONCLUSION

the easiest alternative solution might just be to undo the change in pool filtering and just utilize this existing scope

image

**Comment by yonikid15** [comment]
> ### CONCLUSION
> the easiest alternative solution might just be to undo the change in pool filtering and just utilize this existing scope
> 
> image

Okay, I think I finally figured it out. Thanks for the advice to use that scope. Previously, I thought that using `filterByAvailableInPools` was the solution, since it combines both the `expiryStatus` and the `poolCandidateStatus` in the query. However, I didn't take into account the `orWhereNull('expiry_date)` clause.

**Review by petertgiles — CHANGES_REQUESTED** [review]
Looking really good!  Just a few small things to check.

File: api/app/Models/PoolCandidate.php [hunk 1]
```diff
@@ -357,13 +357,13 @@ public function scopeHasDiploma(Builder $query, bool $hasDiploma): Builder
...
 
-    public function scopeExpiryFilter(Builder $query, ?array $args)
+    public static function scopeExpiryFilter(Builder $query, ?array $args)
     {
         $expiryStatus = isset($args['expiryStatus']) ? $args['expiryStatus'] : ApiEnums::CANDIDATE_EXPIRY_FILTER_ACTIVE;
         if ($expiryStatus == ApiEnums::CANDIDATE_EXPIRY_FILTER_ACTIVE) {
             $query->whereDate('expiry_date', '>=', date("Y-m-d"))
-                ->orWhereNull('expiry_date');
-        } else if ($expiryStatus == ApiEnums::CANDIDATE_EXPIRY_FILTER_EXPIRED) {
+            ->orWhereNull('expiry_date');
```
[the remark this answers is not in the document]
Comment (petertgiles) [reply api/app/Models/PoolCandidate.php:365]: Something like this:
https://stackoverflow.com/a/36372109/18246559

[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/Models/PoolCandidate.php:365]: Thank you Peter 🙌! This solved everything 😆. It took a while to find out why this line was breaking the query. 

This `orWhereNull('expiry_date')` clause was causing the issue in the test `testPoolFilter` where the user was in multiple pools AND that user has at least ONE poolCandidate object that is valid (valid meaning QUALIFIED_AVAILABLE or PLACED_CASUAL and has  EXPRIRY_FILTER_ACTIVE). So, I believe the countPoolCandidatesByPool query would add that user to the other pools `candidateCount` even if the user wasn't valid.

File: api/app/GraphQL/Queries/CountPoolCandidatesByPool.php [hunk 2]
```diff
@@ -78,6 +81,9 @@ public function __invoke($_, array $args)
             }
         });
 
+        // expiry status filter (filter active pool candidates)
+        PoolCandidate::scopeExpiryFilter($queryBuilder, [ 'expiryStatus' => ApiEnums::CANDIDATE_EXPIRY_FILTER_ACTIVE ]);
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/GraphQL/Queries/CountPoolCandidatesByPool.php]: Added back at the top in [f64b12e](https://github.com/GCTC-NTGC/gc-digital-talent/pull/4686/commits/f64b12ef2f757322fe07a57bcbb6628057ce30bf)

File: api/tests/Feature/CountPoolCandidatesByPoolTest.php [hunk 3]
```diff
@@ -735,25 +669,30 @@ public function testSkills()
...
+        PoolCandidate::factory()->create($this->poolCandidateData($pool1, $user));
+
+        // Create a placed casual pool candidate with active expiry date in Pool 2.
+        PoolCandidate::factory()->create($this->poolCandidateData($pool2, $user));
+
+        // Create a screened out pool candidate with active expiry date in Pool 3. Should not appear in response.
+        PoolCandidate::factory()->create($this->poolCandidateData($pool3, $user, false));
+
+        // Create a placed casual pool candidate with inactive expiry date in Pool 3. Should not appear in response.
+        PoolCandidate::factory()->create($this->poolCandidateData($pool3, $user2, true, false));
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/tests/Feature/CountPoolCandidatesByPoolTest.php]: Separate tests added in [80f379d](https://github.com/GCTC-NTGC/gc-digital-talent/pull/4686/commits/80f379d5413d492c62404d68cd8439e7407ce3a3) 👍

File: api/app/Models/PoolCandidate.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/Models/PoolCandidate.php]: Should be fixed by [f64b12e](https://github.com/GCTC-NTGC/gc-digital-talent/pull/4686/commits/f64b12ef2f757322fe07a57bcbb6628057ce30bf)

**Review by petertgiles — APPROVED** [review]
Nice! :1st_place_medal:

**Review by vd1992 — APPROVED** [review]
removing myself as a blocker 😄

**Pull request merged**